### PR TITLE
feat: Loss-based sampling (sample-then-batch only)

### PR DESCRIPTION
**Pull request by MaxiBoether**
This adds a first implementation of sample-then-batch loss-based sampling. Critical follow up tasks are:

- Clean up the interface of the SelectiveSampler. There are many hooks which might not be needed, and some even conflict ("after batch" without the proper actually needed information).
- Generalize the full_finetune recipe to support both sample-then-batch OR batch-then-sample using the same selective sampler. Right now we hardcode sample-then-batch.

I also touch on the `set_num_selected_samples` stuff because it kept giving me trouble and I did not see why it was necessary to do it like that instead of based on the mask.

### CODE REVIEW

File: recipe/full_finetune.py [hunk 1]
```diff
@@ -341,6 +334,21 @@ def setup(self, cfg: DictConfig) -> None:
             (cfg.batch_size, 1), self._loss_fn.ignore_index, device=self._device
         )
 
+    def _set_steps_per_epoch(self) -> None:
```
Comment (Sipondo) [inline recipe/full_finetune.py:339]: This is nice in terms of selection, though I do think this is incompatible with setting up the lr scheduler. We can consider removing LR scheduling altogether if it's not required by our use cases, though it might be nice to support for if we want others to use these DS techniques as an extension to torchtune.

File: recipe/full_finetune.py [hunk 2]
```diff
@@ -673,8 +707,25 @@ def train(self) -> None:
             # in case shuffle is True
             self._sampler.set_epoch(curr_epoch)
 
-            pbar = tqdm(total=self._steps_per_epoch)
             self._sampler.pre_epoch()
+            # ----- First Pass: Scoring Phase -----
```
Comment (Sipondo) [inline recipe/full_finetune.py]: This is a nice solution though I do think it would benefit from being conditional on what the sampler requires. We should make sure the other samplers keep functioning without running an extra forward pass.
Making this conditional also paves the way to bts.

File: recipe/full_finetune.py [hunk 3]
```diff
@@ -673,8 +707,25 @@ def train(self) -> None:
...
+                utils.batch_to_device(batch, self._device)
+                no_grad_mgr = torch.inference_mode() if self._sampler.no_grad_scoring else contextlib.nullcontext()
+                with no_grad_mgr:
+                    logits, shifted_labels = self._forward_pass(batch)
+                    sample_ids = batch["sample_ids"].tolist()
+                    self._sampler.inform_logits(sample_ids, logits, shifted_labels)
+
+            self._sampler.sample()
+            self._set_steps_per_epoch()
+            utils.get_logger("DEBUG").info("Scoring complete; selected %d samples", sum(self._sampler.mask))
```
Comment (Sipondo) [inline recipe/full_finetune.py]: For ease of reading it may be nice to report the amount of resulting batches as well.

File: misc/indexed_ds.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from torch.utils.data import Dataset
+
+class IndexedWrapperDataset(Dataset):
```
Comment (Sipondo) [inline misc/indexed_ds.py:3]: Please add a succinct description (dataset with sample indices added)

File: recipe/full_finetune.py [hunk 5]
```diff
@@ -647,6 +663,24 @@ def _loss_step(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
         del logits
 
         return loss
+    
+    def _forward_pass(self, batch: Dict[str, torch.Tensor]) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (Sipondo) [inline recipe/full_finetune.py:646]: I think it would be better if either this is merged with _loss_step or the name reflects that this is only used for the scoring phase and the names be brought closer together.

File: recipe/full_finetune.py [hunk 6]
```diff
@@ -624,9 +638,11 @@ def save_checkpoint(self, epoch: int) -> None:
             intermediate_checkpoint=(epoch + 1 < self.total_epochs),
         )
 
+    
     def _loss_step(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
         # Shape [b, s], needed for the loss not the model
         labels = batch.pop("labels")
+        sample_ids = batch.pop("sample_ids")
```
Comment (Sipondo) [inline recipe/full_finetune.py]: Please use a throwaway here.

File: selection/selectivesampler.py [hunk 7]
```diff
@@ -50,14 +50,9 @@ def __iter__(self):
         return iter(indices)
 
     def __len__(self) -> int:
```
Comment (Sipondo) [inline selection/selectivesampler.py:52]: Please also remove set_num_selected_samples from Fullsampler.

**Review by Sipondo — CHANGES_REQUESTED** [review]
Thank you for the implementation!
Nothing major, just some things here and there.
Going forward I think BTS and unifying full and lora are on the table, as those two will have some sweeping changes that can affect stuff down the line.

File: recipe/full_finetune.py (hunk 1 above)
Comment (MaxiBoether) [reply recipe/full_finetune.py:339]: Sorry, I don't think I fully follow yet. The code in here was already present in the default recipe. I just factored it into it's own function to run it again after the sampling phase, because one epoch first consists of the forward passes, sampling, then fwd-bwd on subset. Where exactly does lr scheduling break?

File: recipe/full_finetune.py (hunk 3 above)
Comment (MaxiBoether) [reply recipe/full_finetune.py]: That's given in the resulting tqdm (pbar = tqdm(desc="Training", total=self._steps_per_epoch))

File: recipe/full_finetune.py (hunk 2 above)
Comment (MaxiBoether) [reply recipe/full_finetune.py]: The issue here is that in the current stage of development it makes sense to have this as a flag on the sampler but with the next PR the same sampler (e.g., loss sampler) should be used in BtS or StB setting. So this is not really a flag of the sampler, but I also don't want to introduce a StB/BtS flag just yet. Can we make this optional with the next PR?

File: misc/indexed_ds.py (hunk 4 above)
Comment (MaxiBoether) [reply misc/indexed_ds.py:3]: done

File: recipe/full_finetune.py (hunk 5 above)
Comment (MaxiBoether) [reply recipe/full_finetune.py:646]: merged

File: recipe/full_finetune.py (hunk 6 above)
Comment (MaxiBoether) [reply recipe/full_finetune.py]: done

File: selection/selectivesampler.py (hunk 7 above)
Comment (MaxiBoether) [reply selection/selectivesampler.py:52]: done

**Comment by MaxiBoether** [comment]
Addressed comments / replied where I was unsure :)

File: recipe/full_finetune.py (hunk 1 above)
Comment (Sipondo) [reply recipe/full_finetune.py:339]: In setup() the LR scheduler is initialised using steps per epoch, prior to the sampler doing anything.

File: recipe/full_finetune.py (hunk 2 above)
Comment (Sipondo) [reply recipe/full_finetune.py]: Yeah I agree there, though I think it might mean that this merge would break some things.
In any way quickly need to do a follow-up pr with config options, which is where this should live.

File: recipe/full_finetune.py (hunk 3 above)
Comment (Sipondo) [reply recipe/full_finetune.py]: That's fair, in my mind it would've been clearer to report the number separately instead of included in the tqdm but they are close enough together that this is ok as well.

File: recipe/full_finetune.py (hunk 1 above)
Comment (MaxiBoether) [reply recipe/full_finetune.py:339]: Fixed. We just re-init the scheduler after sampling. We don't lose state and there we have access to the actual # batches. in a typical finetuning setup we want to use lr decay since that helps in improving model performance on downstream benchmarks, so it's important we support that

**Review by Sipondo — APPROVED** [review]
